runtime/src/snapshot_utils.rs

```diff
@@ -877,6 +881,99 @@ pub fn create_accounts_run_and_snapshot_dirs(
     Ok((run_path, snapshot_path))
 }
 
+/// Return account path from the appendvec path after checking its format.
+fn get_account_path_from_appendvec_path(appendvec_path: &Path) -> Option<PathBuf> {
+    let run_path = appendvec_path.parent()?;
+    let run_file_name = run_path.file_name()?;
+    // All appendvec files should be under <account_path>/run/.
+    // When generating the bank snapshot directory, they are hardlinked to <account_path>/snapshot/<slot>/
+    if run_file_name != "run" {
+        error!(
+            "The account path {} does not have run/ as its immediate parent directory.",
+            run_path.display()
+        );
+        return None;
+    }
+    let account_path = run_path.parent()?;
+    Some(account_path.to_path_buf())
+}
+
+/// From an appendvec path, derive the snapshot hardlink path.  If the corresponding snapshot hardlink
+/// directory does not exist, create it.
+fn get_snapshot_accounts_hardlink_dir(
+    appendvec_path: &Path,
+    bank_slot: Slot,
+    account_paths: &mut HashSet<PathBuf>,
+    hardlinks_dir: impl AsRef<Path>,
+) -> Result<PathBuf> {
+    let account_path = get_account_path_from_appendvec_path(appendvec_path)
+        .ok_or_else(|| SnapshotError::InvalidAppendVecPath(appendvec_path.to_path_buf()))?;
+
+    let snapshot_hardlink_dir = account_path.join("snapshot").join(bank_slot.to_string());
+
+    // Use the hashset to track, to avoid checking the file system.  Only set up the hardlink directory
+    // and the symlink to it at the first time of seeing the account_path.
+    if !account_paths.contains(&account_path) {
+        let idx = account_paths.len();
+        info!(
+            "for appendvec_path {}, create hard-link path {}",
+            appendvec_path.display(),
+            snapshot_hardlink_dir.display()
+        );
+        fs::create_dir_all(&snapshot_hardlink_dir).map_err(|e| {
+            SnapshotError::IoWithSourceAndFile(
+                e,
+                "create hard-link dir",
+                snapshot_hardlink_dir.clone(),
+            )
+        })?;
+        let symlink_path = hardlinks_dir.as_ref().join(format!("account_path_{idx}"));
+        symlink::symlink_dir(&snapshot_hardlink_dir, symlink_path)?;
```

Comment:
The other errors have mapped errors so that we get the filename in the error message. Should we do that here as well?

Reply:
yes, added that. thanks